libservo: Improve finding python

jschwe opened the pull request:
Servo is built in a virtual environment, which sets [`VIRTUAL_ENV`](https://docs.python.org/3/library/venv.html) to the base path of the venv.
PYTHON3 is only set if mach or the user set it.

Additional changes:

- Use Path / var_os for Paths instead of strings. In General using Path APIs is preferable, since in rare cases valid paths may not be valid utf-8.
- Don't search for python 3.8 anymore, since we require a newer version anyway.
- Don't add the .exe suffix anymore, since Command::new() [will take care of that](https://doc.rust-lang.org/std/process/struct.Command.html#platform-specific-behavior).

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix `cargo build` failing when python3.8 is installed (building without mach)

jschwe committed:
libservo: Improve finding python

Servo is built in a virtual environment, which sets `VIRTUAL_ENV` to
the base path of the venv.
PYTHON3 is only set if mach or the user set it.

Additional changes:

- Use Path / var_os for Paths instead of strings.
 In General using Path APIs is preferable, since in rare cases
 valid paths may not be valid utf-8.
- Don't search for python 3.8 anymore, since
  we require a newer version anyway.
- Don't add the .exe suffix anymore, since
  Command::new() will take care of that.

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>
Signed-off-by: Jonathan Schwender <jonathan.schwender@huawei.com>

Taym95 commented on components/servo/build.rs:
```diff
@@ -23,26 +23,56 @@ fn main() {
+    let mut candidates = vec![];
+    if let Some(venv) = env::var_os("VIRTUAL_ENV") {
+        let mut base = PathBuf::from(venv);
+        base.push("bin");
```
I think you can use `join` to avoid `push`:
```
let base = PathBuf::from(venv).join("bin");
let python = base.join("python");
```

Wuelle commented on components/servo/build.rs:
```diff
@@ -23,26 +23,54 @@ fn main() {
+            .arg("--version")
+            .output()
+            .ok()
+            .map_or(false, |out| out.status.success())
```
Could use `Option::is_some_and` here

sagudev commented:
BTW similar code is also at https://github.com/servo/servo/blob/f1ad364ec2dacca3ec7d79830ef8da9f26fbf4e2/components/script/build.rs#L67
so it would be nice to keep it in sync.

jschwe committed:
script: Improve finding python

Synchronize `find_python` in scripts build.rs with the version in
libservo.

Signed-off-by: Jonathan Schwender <jonathan.schwender@huawei.com>

mrobinson commented on components/script/build.rs:
```diff
@@ -64,26 +64,55 @@ impl<'a> phf_shared::PhfHash for Bytes<'a> {
+/// Tries to find a suitable python
+///
+/// Algorithm
+/// 1. Trying to find python3/python in $VIRTUAL_ENV (this should be from servos venv)
```
```suggestion
/// 1. Trying to find python3/python in $VIRTUAL_ENV (this should be from Servo's venv)
```

mrobinson commented on components/script/build.rs:
```diff
@@ -64,26 +64,55 @@ impl<'a> phf_shared::PhfHash for Bytes<'a> {
+fn find_python() -> PathBuf {
+    let mut candidates = vec![];
+    if let Some(venv) = env::var_os("VIRTUAL_ENV") {
+        let base = PathBuf::from(venv);
+        let bin = base.join("bin");
+        let python = bin.join("python");
+        let python3 = bin.join("python3");
+        if python3.exists() {
```
```suggestion
        let bin_directory = PathBuf::from(venv).join("bin");
        
        let python3 = bin_directory.join("python3");
        if python3.exists() {
```

mrobinson commented on components/script/build.rs:
```diff
@@ -64,26 +64,55 @@ impl<'a> phf_shared::PhfHash for Bytes<'a> {
-            candidates.join(", ")
-        )
-    })
+        if python.exists() {
+            candidates.push(python);
+        }
```
```suggestion
        let python = bin.join("python");
        if python.exists() {
            candidates.push(python);
        }
```

mrobinson commented on components/script/build.rs:
```diff
@@ -64,26 +64,55 @@ impl<'a> phf_shared::PhfHash for Bytes<'a> {
+        .map(|c| c.into_os_string())
+        .collect::<Vec<_>>();
+    panic!(
+        "Can't find python (tried {:?})! Try enabling servos python venv, \
```
```suggestion
        "Can't find python (tried {:?})! Try enabling Servo's Python venv, \
```

mrobinson reviewed (approved):
Looks good with some nits for both versions. It would be nice to eventually share this code somehow.

jschwe commented:
>  It would be nice to eventually share this code somehow.

One could do something like `include!()` or 
```
#[path = ../utils/find_python.rs]
mod find_python;
```

This avoids adding another dependency to the build-script and should thus not influence compile-times.

jschwe committed:
Apply suggestions from code review

Co-authored-by: Martin Robinson <mrobinson@igalia.com>
Signed-off-by: Jonathan Schwender <55576758+jschwe@users.noreply.github.com>
Signed-off-by: Jonathan Schwender <jonathan.schwender@huawei.com>

jschwe committed:
Fix finding venv python on windows

- On Windows the venv scripts and python binaries are in the
  `Scripts` subdirectory instead of `bin`.
- We shouldn't check if the executable in the venv binary dir
  exists, since the actual file could have a `.exe` suffix.
  We don't need to consider the `.exe` suffix for the filename,
  since `Command` will handle that for us.
  We also anyway check the validity of the candidate file, by
  running `$candidate --version`.

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe commented:
windows try build is still running https://github.com/jschwe/servo/actions/runs/10816401959/job/30007519006, but it should pass now. The main issue was that on windows `python` is under `$VIRTUAL_ENV/Scripts` instead of in the `bin` directory as on Linux / macos.
